engine/timer_task_queue.go

```diff
@@ -0,0 +1,295 @@
+	// It is using heap for perf because there could be new timers coming later
+	// The timer will be popped out when it is fired, and sent to the processor.
+	remainingToFireTimersHeap TimerTaskPriorityQueue
+
+	// this tracks the fired timer that are waiting for completed by processor.
+	// The key is the timer sequence. When all timers are completed and no timers in remainingToFireTimersHeap,
+	// it will trigger the next preload when the nextPreloadTimer fires
+	firedToCompleteTimerSequenceMap map[int64]struct{}
+
+	// tasksCompletionChan is the channel to receive completed tasks from processor.
+	// Receiving the completion signal so that it can remove the timer from firedToCompleteTimerSequenceMap
+	// Note that differently from worker task, this timer queue doesn't do batch deletion for "committing".
+	// It relies on the processor to delete the task during processing.
+	tasksCompletionChan chan persistence.TimerTask
+
+	// the channel to receive trigger of polling for newly created timers.
+	// The queue will check if the new timers are within the currentMaxLoadedTaskTimestamp, and if so,
+	// it will load the new timers and add to the remainingToFireTimersHeap
+	triggeredPollingChan chan xdbapi.NotifyTimerTasksRequest
+}
+
+func NewTimerTaskQueueImpl(
+	rootCtx context.Context, shardId int32, cfg config.Config, store persistence.ProcessStore,
+	processor TimerTaskProcessor, logger log.Logger,
+) TimerTaskQueue {
+	qCfg := cfg.AsyncService.TimerTaskQueue
+
+	return &timerTaskQueueImpl{
+		shardId: shardId,
+		store:   store,
+		logger:  logger.WithTags(tag.Shard(shardId)),
+		rootCtx: rootCtx,
+		cfg:     cfg,
+
+		processor: processor,
+
+		nextPreloadTimer: NewLocalTimerGate(logger),
+		nextFiringTimer:  NewLocalTimerGate(logger),
+
+		currMaxLoadedTaskSequence:  0,
+		currMaxLoadedTaskTimestamp: 0,
+
+		remainingToFireTimersHeap:       nil,
+		firedToCompleteTimerSequenceMap: make(map[int64]struct{}),
+		tasksCompletionChan:             make(chan persistence.TimerTask, qCfg.ProcessorBufferSize),
+		triggeredPollingChan:            make(chan xdbapi.NotifyTimerTasksRequest, qCfg.TriggerNotificationBufferSize),
+	}
+}
+
+func (w *timerTaskQueueImpl) Stop(ctx context.Context) error {
+	// close timer to prevent goroutine leakage
+	w.nextPreloadTimer.Close()
+	w.nextFiringTimer.Close()
+
+	return nil
+}
+
+func (w *timerTaskQueueImpl) TriggerPollingTasks(req xdbapi.NotifyTimerTasksRequest) {
+	if req.ShardId != w.shardId {
+		panic("shardId doesn't match")
+	}
+	w.triggeredPollingChan <- req
+}
+
+func (w *timerTaskQueueImpl) Start() error {
+	w.processor.AddTimerTaskQueue(w.shardId, w.tasksCompletionChan)
+
+	w.nextPreloadTimer.Update(time.Now()) // fire immediately to make the first poll
+
+	go func() {
+		for {
+			select {
+			case <-w.nextPreloadTimer.FireChan():
+				if w.shouldLoadNextBatch() {
+					w.loadAndDispatchAndPrepareNext()
+				}
+			case <-w.nextFiringTimer.FireChan():
```

Comment:
Besides firing the heap[0] after loadAndDispatch,  where is the logic to contiously fire other tasks in the heap?

Reply:
in `sendFiredTimerToProcessor()`, there is a loop which continuously pop the item if it's less than Now().